projects/solar/test/test_fan_control_solar.c

```diff
@@ -0,0 +1,234 @@
+  gpio_init();
+  gpio_it_init();
+  data_store_init();
+
+  // Test Board A with 6 MPPTs uses virtual port A
+  GpioAddress s_overtemp_addr = { .port = GPIO_PORT_A, .pin = 0 };
+  GpioAddress s_fan_fail_addr = { .port = GPIO_PORT_A, .pin = 1 };
+  GpioAddress s_full_speed_addr = { .port = GPIO_PORT_A, .pin = 2 };
+
+  // arbitrarily selected value for testing
+  uint16_t s_full_speed_temp_threshold = 3;
+
+  SolarMpptCount s_mppt_count = SOLAR_BOARD_6_MPPTS;
+}
+
+// Used to reinitialize static testing variables Event s_e & GpioState s_gpio_state
+void teardown_test(void) {
+  static GpioState s_gpio_state;
+  static Event s_e;
+}
+
+// Test that "overtemperature" & "fan fail" faults are raised properly
+void test_fan_fail_fault_handling(void) {
+  FanControlSettingsSolar settings = {
+    .overtemp_addr = s_overtemp_addr,
+    .fan_fail_addr = s_fan_fail_addr,
+    .full_speed_addr = s_full_speed_addr,
+    .full_speed_temp_threshold = s_full_speed_temp_threshold,
+    .mppt_count = s_mppt_count,
+  };
+
+  TEST_ASSERT_OK(fan_control_init(&settings));
+
+  // Test that there are no faults initially
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  // Test that fan fail fault is raised correctly
+  // The MAX6643 will call the prv_callback function when fan fail is detected
+  // Parameters have no effect on function
+  prv_fanfail_callback(&s_dummy);
+  event_process(&s_e);
+  MS_TEST_HELPER_ASSERT_EVENT(s_e, EE_SOLAR_FAULT_FAN_FAIL, 0);
+}
+
+// Test that overtemperature fault/pins & full speed pins are raised correctly
+// Case 1: Temperature below threshold for all MPPTs
+void test_fan_overtemp_fault_fullspeed_handling_case_belowtemp(void) {
+  FanControlSettingsSolar settings = {
+    .overtemp_addr = s_overtemp_addr,
+    .fan_fail_addr = s_fan_fail_addr,
+    .full_speed_addr = s_full_speed_addr,
+    .full_speed_temp_threshold = s_full_speed_temp_threshold,
+    .mppt_count = s_mppt_count,
+  };
+
+  // Ensure fan_control_init works
+  TEST_ASSERT_OK(fan_control_init(&settings));
+
+  for (Mppt i = 1; i <= 6; i++) {
+    data_store_set(DATA_POINT_TEMPERATURE(i), 2);
+  }
+  data_store_done();
+
+  // Overtemperature fault should not be raised
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  // Full Speed pin should be high to disable full speed
+  gpio_get_state(&s_full_speed_addr, &s_gpio_state);
+  TEST_ASSERT_EQUAL(GPIO_STATE_HIGH, s_gpio_state);
+}
+
+// Case 2: Temperature above threshold for one MPPT
+void test_fan_overtemp_fault_fullspeed_handling_case_oneabovetemp(void) {
+  FanControlSettingsSolar settings = {
+    .overtemp_addr = s_overtemp_addr,
+    .fan_fail_addr = s_fan_fail_addr,
+    .full_speed_addr = s_full_speed_addr,
+    .full_speed_temp_threshold = s_full_speed_temp_threshold,
+    .mppt_count = s_mppt_count,
+  };
+  TEST_ASSERT_OK(fan_control_init(&settings));
+
+  for (Mppt i = 1; i <= 5; i++) {
+    data_store_set(DATA_POINT_TEMPERATURE(i), 2);
+  }
+  data_store_set(DATA_POINT_TEMPERATURE(6), 4);
```

Comment:
Again the `4` should be abstracted into a `#define ABOVE_THRESHOLD_TEMPERATURE_DC 4` to make it easier to read